### DISCUSSION TITLE
Sources to approximate a dipole in a uLED

### DISCUSSION THREAD

**Discussion by chuanhong98** [facebookresearch/fmmax#114] [Q&A] [answered]
Hello developers and researchers,

thank you for the code and the great paper published. I am currently in the midst of my master's and my professors and I are looking to employ fmmax for inverse designs of uLED and compare the results to FDTD simulations.

When running the code uLED in the examples given, it seems that the selection of FWHM for the gaussian do play a role in the convergence and they affect the extraction efficiency that I obtain when its varied. How should I go ahead in its selection then in relation to the wavelength that I wish to simulate? Is the FWHM tied to the diffraction limit?

I have also noticed that you have provided another source in the source file which is the dirac delta function. Why did you choose a gaussian to approximate a dipole here instead of the dirac delta function and is there a case where they are both identical?

I thank you in advance and look forward to your reply :)

**Comment by mfschubert** [comment, marked as answer]
Hi @chuanhong98, I have also noticed the effect of dipole fwhm on convergence, but didn't go into depth in the paper.

Ideally, you would use a zero-fwhm Gaussian (or equivalently, the dirac delta dipole) as the source. In principle, this has the closest correspondence to the actual physics. However, in the truncated Fourier basis, representing a profile with such rapid spatial variation is not really possible, and so the actual modeled dipole would have finite width *and* ringing, i.e. additional content spatially separated from the dipole location. This can be verified by creating a dipole profile and 

With a Gaussian having finite width, you will obviously still be modeling an extended source, but you can avoid the ringing seen in the delta dipole case. I have seen that this leads to better convergence, and gives results that correspond better to a high-fidelity simulation that includes a narrow dipole and a large number of Fourier terms.

Here is a code snippet that can help explore this effect, and an example figure.

```python
import jax.numpy as jnp
import matplotlib.pyplot as plt
from fmmax import basis, fft, sources

for approximate_num_terms, fwhm in [(1000, 0.01), (1000, 0.025)]:
    primitive_lattice_vectors=basis.LatticeVectors(u=basis.X, v=basis.Y)
    expansion = basis.generate_expansion(
        primitive_lattice_vectors=primitive_lattice_vectors,
        approximate_num_terms=approximate_num_terms,
        truncation=basis.Truncation.CIRCULAR,
    )
    # Place a diploe at x=0, y=0.5
    profile = sources.gaussian_source(
        fwhm=fwhm,
        location=jnp.asarray([[0.0, 0.5]]),
        in_plane_wavevector=jnp.zeros((2,)),
        primitive_lattice_vectors=primitive_lattice_vectors,
        expansion=expansion,
    )

    profile_ifft = fft.ifft(profile, expansion, (1000, 1000), axis=-2)
    profile_ifft /= jnp.amax(jnp.abs(profile_ifft))
    # Plot a cross section at x=0
    plt.semilogy(
        jnp.linspace(0, 1, profile_ifft.shape[1]),
        jnp.abs(profile_ifft)[0, :, 0],
        label=f"N={approximate_num_terms}, fwhm={fwhm:.3f}"
    )

plt.legend()
plt.xlabel("y")
```
![image](https://github.com/facebookresearch/fmmax/assets/30735893/2ec57a14-2c3d-4960-830f-8354f5736694)

To ultimately settle on an fwhm value, it is probably best to perform a convergence study, and then simply pick the value that is sufficiently converged for your purposes. I am not sure there is a much more rigorous solution here.

One final point to emphasize is that the simulation settings used during the optimization itself need not be the same as those used for convergence check. In fact, it is probably best to use settings that result in better convergence, to avoid the overfitting effect we showed in the paper.

**Reply by chuanhong98** [reply]
Hi Martin, 

thank you for your swift answer and the snippet of the code provided. I understand what you mean now and I have since then tried to replicate the structure and the results you have published in the paper using the uLED file as a basis.

I have managed to replicate the results and also simulated the results with a lower FWHM Gaussian of 20nm and a Dirac-Delta Function and the results are as such: 
![328192919-60ba2245-ca54-4fa0-9d36-e8328eef295c](https://github.com/facebookresearch/fmmax/assets/168533583/6c5f68f5-fa70-4565-8bf4-f154cf50c56b)

I have found that a Gaussian with a larger FWHM leads to a higher light extraction efficiency and for the dirac-delta dipole, even for Fourier Terms of around N=5500 the light extraction efficiency does not seem to have shown convergence. 

Is there perhaps a basis on why you have chosen a Gaussian with FWHM of 60nm specifically for the simulation with FMMAX and FDTD?

I thank you in advance and look forward to hearing back from you.

**Reply by mfschubert** [reply]
Hi @chuanhong98, I did not see such problems with convergence for small fwhm, and the choice of 60 nm was a bit arbitrary.

One detail that is mentioned in the paper (section 3), and may be important here, is that the emitted power is measured in planes 50 nm above and below the dipole. The simple uled example doesn't seem to do this, and instead measures power directly at the dipole plane. You may try to adjust your model in the same way, to see if this improves things.

**Reply by chuanhong98** [reply]
hi @mfschubert, after some tinkering around I think I finally got the results and understand more about the code and I just wanted to confirm about my thinking and thought processes. Thank you for the tips and directions pointed. 

![output](https://github.com/facebookresearch/fmmax/assets/168533583/fa4dabd2-4b99-4104-adaf-4425732462a2)

These are my results obtained after adding 2 additional 50nm planes before and after the dipole and the convergence obtained is brilliant. 

What I did was basically so: added 2 additional planes, used fields.propagate_amplitude for forward and backward fields to the top and bottom planes that I wanted, generated the flux and did the subtraction, following the examples given in the uLED file.

![Capture](https://github.com/facebookresearch/fmmax/assets/168533583/5b595959-6656-4922-b6c8-30d568f12c5e)

I thank you in advance and look forward to your reply :) have a nice weekend

**Reply by mfschubert** [reply]
Nice work, everything looks good and matches what I have previously observed.
